opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java

```diff
@@ -0,0 +1,111 @@
+
+import io.opentracing.NoopTracerFactory;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.propagation.Format;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Forwards all methods to another tracer that can be configured by calling {@link #register(Tracer)}.
+ * <p>
+ * When the tracer is needed it delegates all tracing methods to
+ * either the <em>registered</em> global tracer
+ * or the {@link io.opentracing.NoopTracer NoopTracer}.
+ */
+public final class GlobalTracer implements Tracer {
+    private static final Logger LOGGER = Logger.getLogger(GlobalTracer.class.getName());
+
+    /**
+     * Singleton instance.
+     * <p>
+     * Since we cannot prevent people using {@linkplain #get() GlobalTracer.get()} as a constant,
+     * this guarantees that references obtained before, during or after initialization
+     * all behave as if obtained <em>after</em> initialization once properly initialized.<br>
+     * As a minor additional benefit it makes it harder to circumvent the {@link Tracer} API.
+     */
+    private static final GlobalTracer INSTANCE = new GlobalTracer();
+
+    /**
+     * The resolved {@link Tracer} to delegate to, or {@code null} if none was registered yet.
+     */
+    private static volatile Tracer globalTracer = null;
+
+    private GlobalTracer() {
+    }
+
+    /**
+     * Returns the constant {@linkplain GlobalTracer}.
+     * <p>
+     * All methods are forwarded to the currently configured tracer.<br>
+     * Until a tracer is {@link #register(Tracer) explicitly configured},
+     * the {@link io.opentracing.NoopTracer NoopTracer} is used.
+     *
+     * @return The global tracer constant.
+     * @see #register(Tracer)
+     */
+    public static Tracer get() {
+        return INSTANCE;
+    }
+
+    /**
+     * Explicitly configures a {@link Tracer} to back the behaviour of the {@link #get() global tracer}.
+     * <p>
+     * Registration is a one-time operation, attempting to call it more often will result in a runtime exception.
+     * <p>
+     * Every application intending to use the global tracer is responsible for registering it once
+     * during its initialization.
+     *
+     * @param tracer Tracer to use as global tracer.
+     * @return The previous global tracer or <code>null</code> if there was none.
+     */
+    public static void register(final Tracer tracer) {
+        if (tracer instanceof GlobalTracer) {
+            LOGGER.log(Level.FINE, "Attempted to register the GlobalTracer as delegate of itself.");
+            return; // no-op
+        } else if (globalTracer != null && !globalTracer.equals(tracer)) {
+            throw new IllegalStateException("There is already a current globalTracer registered.");
+        }
+        globalTracer = tracer;
+    }
+
+    private static Tracer globalOrNoopTracer() {
+        return globalTracer == null ? NoopTracerFactory.create() : globalTracer;
```

Comment:
why do we re-create() the NoopTracer every time this is called? seems to me that we could do that work once at static init time.